internal/pkg/config/matcher_test.go

```diff
@@ -0,0 +1,369 @@
+			yaml: `
+Package: foo
+PackageRE: bar`,
+			shouldErrorOnLoad: true,
+		},
+		{
+			desc: "Do not permit both Type and TypeRE",
+			yaml: `
+Type: foo
+TypeRE: bar`,
+			shouldErrorOnLoad: true,
+		},
+		{
+			desc: "Do not permit both Field and FieldRE",
+			yaml: `
+Field: foo
+FieldRE: bar`,
+			shouldErrorOnLoad: true,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			sm := sourceMatcher{}
+			err := yaml.UnmarshalStrict([]byte(tc.yaml), &sm)
+
+			if (err != nil) != tc.shouldErrorOnLoad {
+				t.Errorf("error expectation = %v, but got err=%v", tc.shouldErrorOnLoad, err)
+			}
+		})
+	}
+}
+
+func TestSourceMatcherMatching(t *testing.T) {
+	testCases := []struct {
+		desc, yaml           string
+		path, typ, fieldName string
+		shouldMatchType      bool
+		shouldMatchField     bool
+	}{
+		{
+			desc: "Literal foo.bar (no field args) should match foo.bar and foo.bar.baz",
+			yaml: `
+Package: foo
+Type: bar`,
+			path:             "foo",
+			typ:              "bar",
+			fieldName:        "baz",
+			shouldMatchType:  true,
+			shouldMatchField: true,
+		},
+		{
+			desc: "Literal foo.bar (no field args) should NOT match foodstuff.bar or foodstuff.bar.baz",
+			yaml: `
+Package: foo
+Type: bar`,
+			path:             "foodstuff",
+			typ:              "bar",
+			fieldName:        "baz",
+			shouldMatchType:  false,
+			shouldMatchField: false,
+		},
+		{
+			desc: "Regexp foo.bar (no field args) should match foodstuff.bar and foodstuff.bar.baz",
+			yaml: `
+PackageRE: foo
+TypeRE: bar`,
+			path:             "foodstuff",
+			typ:              "bar",
+			fieldName:        "baz",
+			shouldMatchType:  true,
+			shouldMatchField: true,
+		},
+		{
+			desc: "Mixed regexp and literal matchers are permitted",
+			yaml: `
+PackageRE: foo
+Type: bar
+Field: baz`,
+			path:             "foodstuff",
+			typ:              "bar",
+			fieldName:        "baz",
+			shouldMatchType:  true,
+			shouldMatchField: true,
+		},
+		{
+			desc: "Literal foo.bar.qux should MatchType foo.bar but not MatchField foo.bar.baz",
+			yaml: `
+PackageRE: foo
+Type: bar
+Field: qux`,
+			path:             "foodstuff",
+			typ:              "bar",
+			fieldName:        "baz",
+			shouldMatchType:  true,
+			shouldMatchField: false,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			sm := sourceMatcher{}
+			if err := yaml.UnmarshalStrict([]byte(tc.yaml), &sm); err != nil {
+				t.Errorf("Unexpected error unmarshalling sourceMatcher: %v", err)
+			}
+
+			if tc.shouldMatchType != sm.MatchType(tc.path, tc.typ) {
+				t.Errorf("MatchType(%q, %q) = %v; got %v", tc.path, tc.typ, tc.shouldMatchType, !tc.shouldMatchType)
+			}
+			if tc.shouldMatchField != sm.MatchField(tc.path, tc.typ, tc.fieldName) {
+				t.Errorf("MatchField(%q, %q, %q) = %v; got %v", tc.path, tc.typ, tc.fieldName, tc.shouldMatchType, !tc.shouldMatchType)
+			}
+		})
+	}
+}
+
+func TestMatcherTypes(t *testing.T) {
```

Comment:
Nit: I think each matcher type should have its own test, but I don't think it's worth changing.